sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java

```diff
@@ -42,15 +42,12 @@
    *                    Usually Spark processes many RDD partitions at the same time,
    *                    implementations should use the partition id to distinguish writers for
    *                    different partitions.
-   * @param attemptNumber Spark may launch multiple tasks with the same task id. For example, a task
-   *                      failed, Spark launches a new task wth the same task id but different
-   *                      attempt number. Or a task is too slow, Spark launches new tasks wth the
-   *                      same task id but different attempt number, which means there are multiple
-   *                      tasks with the same task id running at the same time. Implementations can
-   *                      use this attempt number to distinguish writers of different task attempts.
+   * @param taskId A unique identifier for a task that is performing the write of the partition
+   *               data. Spark may run multiple tasks for the same partition (due to speculation
+   *               or task failures, for example).
    * @param epochId A monotonically increasing id for streaming queries that are split in to
    *                discrete periods of execution. For non-streaming queries,
    *                this ID will always be 0.
    */
-  DataWriter<T> createDataWriter(int partitionId, int attemptNumber, long epochId);
+  DataWriter<T> createDataWriter(int partitionId, int taskId, long epochId);
```

Comment:
in `SparkHadoopWriter` we must have a int,  but here why not just change the type to long? data source v2 is still evolving and we already made a lot of changes in the master branch.

Reply:
I'm find with that if you're ok with it, but wouldn't that make backporting this to 2.3 a little fishy? Yeah it's evolving, but it's still a little sub-optimal to break things in a maintenance release.

Reply:
if this patch targets 2.3, I'd say we should not change any API or document, just pass `taskId.toInt` as `attemptNumber` and add comments to explain this hacky workaround.

Reply:
Just so I understand, what's the reason for not changing the parameter name and API docs? The name is not a public API in Java, so it doesn't break anything.

And regardless of the parameter name, the API documentation is wrong (since it says you can have multiple tasks with the same ID, but different attempts, which does not happen).

Reply:
The V2 commit stuff is not in 2.3

Reply:
https://issues.apache.org/jira/browse/SPARK-23323

Reply:
Hmm, interesting. But there is an API in 2.3:

```
DataWriter<T> createDataWriter(int partitionId, int attemptNumber);
```

Which I guess would still suffer from the problem Ryan describes in the bug. In any case, that makes it not possible to cleanly backport this, so we can make the type change here.

Reply:
+1 for the type change.